### PR TITLE
Implement download-bundle, upload-bundle commands to support full migration of plugins

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

- This PR implements `tanzu plugin download-bundle` and `tanzu plugin upload-bundle` commands
  - Note: Currently these commands only support the entire plugin inventory image migration. Partial migration will be added as a separate PR for easy code review.

_Test Central Repository Changes:_
- This PR adds 2 more test repository `localhost:9876/tanzu-cli/plugins/airgapped:small` and `localhost:9876/tanzu-cli/plugins/airgapped:large` which contains all plugin stubs for all plugins and all available versions of those plugins.

_Additional changes:_

- Updates `go.mod` to use the latest (`v0.36.1`) version of `imgpkg` (`github.com/vmware-tanzu/carvel-imgpkg v0.36.1`)
  - Previously used version was `v0.17.0`
- Converts the image operations available at (https://github.com/vmware-tanzu/tanzu-cli/blob/main/pkg/carvelhelpers/fetcher.go) to interface implementation.
  - This should allow us to write unit tests efficiently
- Adds unit tests for bundle download and upload

### Which issue(s) this PR fixes

Related to #223

### Describe testing done for PR

- _Build and configure the environment variables_
```
$ make build

# Configure the environment variables to point to 'localhost:9876/tanzu-cli/plugins/airgapped:small' image
$ export TANZU_CLI_PRE_RELEASE_REPO_IMAGE=localhost:9876/tanzu-cli/plugins/airgapped:small
$ export TANZU_CLI_PLUGIN_DISCOVERY_IMAGE_SIGNATURE_VERIFICATION_SKIP_LIST=localhost:9876/tanzu-cli/plugins/airgapped:small
```

- _Verify the setup with search command and check all available plugins_
```
$ tz plugin search
[i] Reading plugin inventory for "localhost:9876/tanzu-cli/plugins/airgapped:small", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:9876/tanzu-cli/plugins/airgapped:small"
 
  NAME              DESCRIPTION                TARGET           LATEST  
  isolated-cluster  Desc for isolated-cluster  global           v0.0.1  
  cluster           Desc for cluster           kubernetes       v0.0.1  
  account           Desc for account           mission-control  v0.0.1  

$ tz plugin search --name account --show-details
name: account
description: Desc for account
target: mission-control
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9

$ tz plugin search --name cluster --show-details
name: cluster
description: Desc for cluster
target: kubernetes
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9

$ tz plugin search --name isolated-cluster --show-details
name: isolated-cluster
description: Desc for isolated-cluster
target: global
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9
```

- Run the download-bundle command
```
$ tz plugin download-bundle --image localhost:9876/tanzu-cli/plugins/airgapped:small --to-tar /tmp/2/plugin_bundle.tar
[i] downloading image "localhost:9876/tanzu-cli/plugins/airgapped:small"
[i] copy | exporting 1 images...
[i] copy | will export localhost:9876/tanzu-cli/plugins/airgapped@sha256:bfd52cae3bcdadfc868331bb232e4431c9a752ccdd089edcf7df3afdb058a384
[i] copy | exported 1 images
[i] copy | writing layers...
[i] copy | done: file 'manifest.json' (16.628µs)
[i] copy | done: file 'sha256-e78ea77ef930e968f8e9623dc786753677db445d066d7bd7dad3ec9dc2f020e6.tar.gz' (77.864µs)
[i] ---------------------------
[i] downloading image "localhost:9876/tanzu-cli/plugins/vmware/tkg/darwin/amd64/mission-control/account:v0.0.1"
[i] copy | exporting 1 images...
[i] copy | will export localhost:9876/tanzu-cli/plugins/vmware/tkg/darwin/amd64/mission-control/account@sha256:c5b908dbebbb4683bab037671c4a30cc09d599a864567460719f3b662cb1a447
[i] copy | exported 1 images
[i] copy | writing layers...
[i] copy | done: file 'manifest.json' (16.213µs)
[i] copy | done: file 'sha256-b5b03d3d065bea842c60693562632b15f7ba772a866e9e4c917bcc25df468bad.tar.gz' (62.699µs)
[i] ---------------------------
.
.
.
[i] saving plugin bundle at: /tmp/2/plugin_bundle.tar
```

- Tar file is available at `/tmp/2/plugin_bundle.tar` manually untar to see the content of the file
```
$ cd /tmp/2
$ ls -lh
total 240
-rw-r--r--  1 anujc  wheel   120K Apr 18 23:53 plugin_bundle.tar

# untar the bundle
$ tar -xf plugin_bundle.tar

# check the directory content
$ tree
.
├── plugin_bundle
│   ├── account-mission-control-darwin_amd64-v0.0.1.tar
│   ├── account-mission-control-darwin_amd64-v9.9.9.tar
│   ├── account-mission-control-darwin_arm64-v0.0.1.tar
│   ├── account-mission-control-darwin_arm64-v9.9.9.tar
│   ├── account-mission-control-linux_amd64-v0.0.1.tar
│   ├── account-mission-control-linux_amd64-v9.9.9.tar
│   ├── account-mission-control-windows_amd64-v0.0.1.tar
│   ├── account-mission-control-windows_amd64-v9.9.9.tar
│   ├── cluster-kubernetes-darwin_amd64-v0.0.1.tar
│   ├── cluster-kubernetes-darwin_amd64-v9.9.9.tar
│   ├── cluster-kubernetes-darwin_arm64-v0.0.1.tar
│   ├── cluster-kubernetes-darwin_arm64-v9.9.9.tar
│   ├── cluster-kubernetes-linux_amd64-v0.0.1.tar
│   ├── cluster-kubernetes-linux_amd64-v9.9.9.tar
│   ├── cluster-kubernetes-windows_amd64-v0.0.1.tar
│   ├── cluster-kubernetes-windows_amd64-v9.9.9.tar
│   ├── isolated-cluster-global-darwin_amd64-v0.0.1.tar
│   ├── isolated-cluster-global-darwin_amd64-v9.9.9.tar
│   ├── isolated-cluster-global-darwin_arm64-v0.0.1.tar
│   ├── isolated-cluster-global-darwin_arm64-v9.9.9.tar
│   ├── isolated-cluster-global-linux_amd64-v0.0.1.tar
│   ├── isolated-cluster-global-linux_amd64-v9.9.9.tar
│   ├── isolated-cluster-global-windows_amd64-v0.0.1.tar
│   ├── isolated-cluster-global-windows_amd64-v9.9.9.tar
│   ├── plugin-inventory-image.tar
│   └── plugin_bundle_manifest.yaml
└── plugin_bundle.tar

# check the plugin_bundle_manifest.yaml
$ cat plugin_bundle/plugin_bundle_manifest.yaml 
images:
    - filePath: plugin-inventory-image.tar
      imagePath: /airgapped
    - filePath: account-mission-control-darwin_amd64-v0.0.1.tar
      imagePath: /vmware/tkg/darwin/amd64/mission-control/account
    - filePath: account-mission-control-darwin_arm64-v0.0.1.tar
      imagePath: /vmware/tkg/darwin/arm64/mission-control/account
    - filePath: account-mission-control-linux_amd64-v0.0.1.tar
      imagePath: /vmware/tkg/linux/amd64/mission-control/account
    - <more>
```

- Use the same downloaded tar file to upload it to a different repository (`localhost:9876/test/airgapped/v4/tanzu-cli/plugins/`)
```
$ tz plugin upload-bundle --tar /tmp/2/plugin_bundle.tar --to-repo localhost:9876/test/airgapped/v4/tanzu-cli/plugins/
[i] ---------------------------
[i] uploading image "localhost:9876/test/airgapped/v4/tanzu-cli/plugins/airgapped"
[i] copy | importing 1 images...

2.95 KiB / 2.95 KiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 33.91 KiB p/s
[i] copy | 
copy | done uploading images
[i] copy | Tagging images
[i] ---------------------------
[i] uploading image "localhost:9876/test/airgapped/v4/tanzu-cli/plugins/vmware/tkg/darwin/amd64/mission-control/account"
[i] copy | importing 1 images...

1.12 KiB / 1.12 KiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 13.14 KiB p/s
[i] copy | 
copy | done uploading images
[i] copy | Tagging images
[i] ---------------------------
[i] uploading image "localhost:9876/test/airgapped/v4/tanzu-cli/plugins/vmware/tkg/darwin/arm64/mission-control/account"
[i] copy | importing 1 images...
.
.
.
```

- Use the new repository with tanzu-cli, discover and install plugins from the new repository (`localhost:9876/test/airgapped/v4/tanzu-cli/plugins/`)

```
# remove all cached databases to make sure cli doesn't use cached database

$ rm -rf ~/.cache/tanzu/plugin_inventory/
$ export TANZU_CLI_PRE_RELEASE_REPO_IMAGE=localhost:9876/test/airgapped/v4/tanzu-cli/plugins/airgapped:small
$ export TANZU_CLI_PLUGIN_DISCOVERY_IMAGE_SIGNATURE_VERIFICATION_SKIP_LIST=localhost:9876/test/airgapped/v4/tanzu-cli/plugins/airgapped:small


$ tz plugin search
[i] Reading plugin inventory for "localhost:9876/test/airgapped/v4/tanzu-cli/plugins/airgapped:small", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:9876/test/airgapped/v4/tanzu-cli/plugins/airgapped:small"
 
  NAME              DESCRIPTION                TARGET           LATEST  
  isolated-cluster  Desc for isolated-cluster  global           v0.0.1  
  cluster           Desc for cluster           kubernetes       v0.0.1  
  account           Desc for account           mission-control  v0.0.1  

$ tz plugin search --name isolated-cluster --show-details
name: isolated-cluster
description: Desc for isolated-cluster
target: global
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9

$ tz plugin search --name cluster --show-details
name: cluster
description: Desc for cluster
target: kubernetes
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9

$ tz plugin search --name account --show-details
name: account
description: Desc for account
target: mission-control
latest: v0.0.1
versions:
    - v0.0.1
    - v9.9.9

$ tz plugin install isolated-cluster
[i] Installing plugin 'isolated-cluster:v0.0.1' with target 'global'
[ok] successfully installed 'isolated-cluster' plugin

$ tz plugin install isolated-cluster --version v9.9.9
[i] Installing plugin 'isolated-cluster:v9.9.9' with target 'global'
[ok] successfully installed 'isolated-cluster' plugin

$ tz plugin install cluster --version v0.0.1
[i] Installing plugin 'cluster:v0.0.1' with target 'kubernetes'
[ok] successfully installed 'cluster' plugin

$ tz plugin install cluster --version v9.9.9
[i] Installing plugin 'cluster:v9.9.9' with target 'kubernetes'
[ok] successfully installed 'cluster' plugin

$ tz plugin install account
[i] Installing plugin 'account:v0.0.1' with target 'mission-control'
[ok] successfully installed 'account' plugin

$ tz plugin install account --version v9.9.9
[i] Installing plugin 'account:v9.9.9' with target 'mission-control'
[ok] successfully installed 'account' plugin
```

### Release note

```release-note
Implement download-bundle, and upload-bundle commands for plugin migration to the custom registry
```

### CODE REVIEW

**Comment by marckhouzam** [comment]
Going over this, which looks great up to now!
Just a note unrelated.  
In the PR description you give multiple `tz plugin search --name <plugin> --show-details` commands to check the plugin versions.
I wanted to mention that we support `tz plugin search --show-details` to show details of all plugins at once.

I mentioning this in case you don't think we _should_ support that form of `--show-details` and you think we should remove it.

**Review by marckhouzam — COMMENTED** [review]
Super nice!
Some minor suggestions in-line.

One question: 
Do we want to verify that the image used by the `tz plugin download-bundle` is properly signed _before_ allowing the creating of the airgap repo?  With the same requirement to use the SKIP variable if the user is comfortable with an unsigned origin repo. 

This is a rare case, but it could help the operator creating the repo to notice a security problem, instead of the problem arising when users start using the air-gap repo.

I'm not sure that is really that important, but if we want to do that, I'm fine having it done later on in a separate PR.

Thanks for also upgrading `imgpkg`, we were on a really old version!

File: pkg/airgapped/plugin_bundle_download.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+	// Save all downloaded images as part of manifest file
+	err = saveManifestFile(allImages, tempDir)
+	if err != nil {
+		return errors.Wrap(err, "error while saving plugin bundle manifest")
+	}
+
+	log.Infof("saving plugin bundle at: %s", o.ToTar)
+	// Save entire plugin bundle as a single tar file
+	err = tarinator.Tarinate([]string{tempDir}, o.ToTar)
```
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/airgapped/plugin_bundle_download.go:78]: Thanks for catching it. Added the validation and a unit test.

File: hack/central-repo/upload-plugins.sh [hunk 2]
```diff
@@ -254,7 +262,72 @@ for name in ${tmcPlugins[*]}; do
...
+echo "======================================"
+# Reset the DB
+rm -f $database
+touch $database
+# Create the DB table
+cat $ROOT_DIR/create_tables.sql | sqlite3 -batch $database
+
+for name in ${globalPlugins[0]}; do
+    addPlugin $name global true v0.0.1
+    addPlugin $name global true v9.9.9
```
[the remark this answers is not in the document]
Comment (anujc25) [reply hack/central-repo/upload-plugins.sh:277]: Yeah. Makes sense. I updated the `addPlugin` function under `upload-plugin.sh` to not configure recommended version.

**Comment by anujc25** [comment]
> Do we want to verify that the image used by the tz plugin download-bundle is properly signed before allowing the creating of the airgap repo? With the same requirement to use the SKIP variable if the user is comfortable with an unsigned origin repo.
>
> This is a rare case, but it could help the operator creating the repo to notice a security problem, instead of the problem arising when users start using the air-gap repo.
>
> I'm not sure that is really that important, but if we want to do that, I'm fine having it done later on in a separate PR.

Thanks for bringing this to our attention. I think you brought up a really good point and I also feel we should do signature validation before downloading the bundle. I will file a separate issue to do this.

**Review by marckhouzam — APPROVED** [review]
LGTM
Very nice work.

**Pull request merged**